File: app-rater/src/main/java/com/simplaapliko/apprater/AppRater.java
/*
 * Copyright (C) 2015 Oleg Kan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.simplaapliko.apprater;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.res.Resources;
import android.net.Uri;
import android.widget.Toast;

import java.util.Date;
import java.util.List;

import androidx.fragment.app.FragmentActivity;

public final class AppRater {

    private static final int DAY = 24 * 60 * 60 * 1000;

    /**
     * Check if 'rate app' dialog needs to be displayed,
     * based on the first launch date, launch count and do not show again
     * shared preferences.
     *
     * @param activity FragmentActivity
     */
    public static void appLaunched(FragmentActivity activity) {
        appLaunched(activity, null, null, null);
    }

    /**
     * Check if 'rate app' dialog needs to be displayed,
     * based on the first launch date, launch count and do not show again
     * shared preferences.
     *
     * @param activity FragmentActivity
     * @param onPositiveButtonListener Positive button click listener
     * @param onNegativeButtonListener Negative button click listener
     * @param onNeutralButtonListener Neutral button click listener
     */
    public static void appLaunched(
            FragmentActivity activity,
            DialogInterface.OnClickListener onPositiveButtonListener,
            DialogInterface.OnClickListener onNegativeButtonListener,
            DialogInterface.OnClickListener onNeutralButtonListener) {

        PreferencesHelper preferences = new PreferencesHelper(activity);

        if (preferences.isDoNoShowAgain()) {
            return;
        }

        increment(activity);

        if (isTimeToRate(activity)) {
            showDialog(activity, onPositiveButtonListener, onNegativeButtonListener,
                    onNeutralButtonListener);
        }
    }

    public static void showDialog(FragmentActivity activity) {
        showDialog(activity, null, null, null);
    }

    public static void showDialog(
            FragmentActivity activity,
            DialogInterface.OnClickListener onPositiveButtonListener,
            DialogInterface.OnClickListener onNegativeButtonListener,
            DialogInterface.OnClickListener onNeutralButtonListener) {

        RateAppDialog dialog = new RateAppDialog.Builder()
                .build();

        if (onPositiveButtonListener != null) {
            dialog.setOnPositiveButtonListener(onPositiveButtonListener);
        }

        if (onNegativeButtonListener != null) {
            dialog.setOnNegativeButtonListener(onNegativeButtonListener);
        }

        if (onNeutralButtonListener != null) {
            dialog.setOnNeutralButtonListener(onNeutralButtonListener);
        }

        dialog.show((activity).getSupportFragmentManager(), null);
    }

    /**
     * Increment launch count and set first launch date
     * @param context Context
     */
    public static void increment(Context context) {
        PreferencesHelper preferences = new PreferencesHelper(context);

        // Increment the launch counter
        int launchCount = preferences.getLaunchCount() + 1;
        preferences.setLaunchCount(launchCount);

        // Get date of the first launch
        long firstLaunchDate = preferences.getFirstLaunchDate();
        if (firstLaunchDate == 0) {
            preferences.setFirstLaunchDate(System.currentTimeMillis());
        }
    }

    public static Date getFirstLaunchDate(Context context) {
        PreferencesHelper preferences = new PreferencesHelper(context);
        long firstLaunchDate = preferences.getFirstLaunchDate();
        return new Date(firstLaunchDate);
    }

    public static int getLaunchCount(Context context) {
        PreferencesHelper preferences = new PreferencesHelper(context);
        return preferences.getLaunchCount();
    }

    public static boolean isDoNotShowAgain(Context context) {
        PreferencesHelper preferences = new PreferencesHelper(context);
        return preferences.isDoNoShowAgain();
    }

    public static boolean isTimeToRate(Context context) {
        PreferencesHelper preferences = new PreferencesHelper(context);
        Resources resources = context.getResources();

        int dayUntilPrompt = resources.getInteger(R.integer.ar_app_rater_days_until_prompt);
        int launchesUntilPrompt = resources.getInteger(R.integer.ar_app_rater_launches_until_prompt);
        int launchCount = preferences.getLaunchCount();
        long firstLaunchDate = preferences.getFirstLaunchDate();

        return !preferences.isDoNoShowAgain() &&
                (launchCount >= launchesUntilPrompt &&
                (System.currentTimeMillis() >= firstLaunchDate + dayUntilPrompt * DAY));
    }

    /**
     * Open Play Store for the related application.
     */
    public static void rateApp(Context context) {

        new PreferencesHelper(context)
                .setDoNoShowAgain(true);

        // create Play Store intent
        Intent rate = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + context.getPackageName()));

        if (!isIntentCallable(context, rate)) {
            // if Play Store doesn't exist show in browser
            rate = new Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=" + context.getPackageName()));
        }

        if (!isIntentCallable(context, rate)) {
            // if browser app doesn't exist, show toast and quit
            Toast.makeText(context, context.getString(R.string.ar_unable_to_find_google_play), Toast.LENGTH_LONG).show();
            return;
        }

        if (context instanceof Activity) {
            context.startActivity(rate);
        } else {
            Intent intent = Intent.createChooser(rate, null);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        }
    }

    /**
     * Reset launch counter.
     */
    public static void remindLater(Context context) {
        new PreferencesHelper(context)
                .setLaunchCount(0)
                .setFirstLaunchDate(System.currentTimeMillis());
    }

    /**
     * Cancel reminders.
     */
    public static void cancelReminders(Context context) {
        new PreferencesHelper(context)
                .setDoNoShowAgain(true);
    }

    /**
     * Check if this is a first launch
     */
    public static boolean isFirstLaunch(Context context) {
        return new PreferencesHelper(context)
                .getFirstLaunchDate() == 0;
    }

    private static boolean isIntentCallable(Context context, Intent intent) {
        List<ResolveInfo> list = context.getPackageManager().queryIntentActivities(
                intent,
                PackageManager.MATCH_DEFAULT_ONLY);
        return list.size() > 0;
    }
}

File: app-rater/src/main/java/com/simplaapliko/apprater/AppRater.kt
/*
 * Copyright (C) 2015 Oleg Kan
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.simplaapliko.apprater

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.widget.Toast
import java.util.Date

@Suppress("MemberVisibilityCanBePrivate")
object AppRater {
    private const val DAY = 24 * 60 * 60 * 1000

    /**
     * Check if 'rate app' dialog needs to be displayed,
     * based on the first launch date, launch count and do not show again
     * shared preferences.
     *
     * @param activity Activity
     * @param onPositiveButtonListener Positive button click listener
     * @param onNegativeButtonListener Negative button click listener
     * @param onNeutralButtonListener Neutral button click listener
     */
    @JvmStatic
    @JvmOverloads
    fun appLaunched(
        activity: Activity,
        onPositiveButtonListener: DialogInterface.OnClickListener? = null,
        onNegativeButtonListener: DialogInterface.OnClickListener? = null,
        onNeutralButtonListener: DialogInterface.OnClickListener? = null
    ) {
        val preferences = PreferencesHelper(activity)
        if (preferences.isDoNoShowAgain) {
            return
        }

        increment(activity)

        if (isTimeToRate(activity)) {
            showDialog(activity, onPositiveButtonListener, onNegativeButtonListener, onNeutralButtonListener)
        }
    }

    @JvmStatic
    @JvmOverloads
    fun showDialog(
        activity: Activity,
        onPositiveButtonListener: DialogInterface.OnClickListener? = null,
        onNegativeButtonListener: DialogInterface.OnClickListener? = null,
        onNeutralButtonListener: DialogInterface.OnClickListener? = null
    ) {
        val builder = AlertDialog.Builder(activity)
            .setCancelable(true)
            .setTitle(R.string.ar_dialog_rate_title)
            .setMessage(R.string.ar_dialog_rate_message)

        onPositiveButtonListener?.let { listener ->
            builder.setPositiveButton(
                R.string.ar_dialog_rate_positive_button_rate
            ) { dialog: DialogInterface?, which: Int ->
                rateApp(activity)
                listener.onClick(dialog, which)
            }
        }

        onNegativeButtonListener?.let { listener ->
            builder.setNegativeButton(
                R.string.ar_dialog_rate_negative_button
            ) { dialog: DialogInterface?, which: Int ->
                remindLater(activity)
                listener.onClick(dialog, which)
            }
        }

        onNeutralButtonListener?.let { listener ->
            builder.setNeutralButton(
                R.string.ar_dialog_rate_neutral_button
            ) { dialog: DialogInterface?, which: Int ->
                cancelReminders(activity)
                listener.onClick(dialog, which)
            }
        }

        builder.create()
            .show()
    }

    /**
     * Increment launch count and set first launch date
     * @param context Context
     */
    fun increment(context: Context) {
        val preferences = PreferencesHelper(context)

        // Increment the launch counter
        val launchCount = preferences.launchCount + 1
        preferences.setLaunchCount(launchCount)

        // Get date of the first launch
        val firstLaunchDate = preferences.firstLaunchDate
        if (firstLaunchDate == 0L) {
            preferences.setFirstLaunchDate(System.currentTimeMillis())
        }
    }

    @JvmStatic
    fun getFirstLaunchDate(context: Context): Date {
        val preferences = PreferencesHelper(context)
        val firstLaunchDate = preferences.firstLaunchDate
        return Date(firstLaunchDate)
    }

    @JvmStatic
    fun getLaunchCount(context: Context): Int {
        val preferences = PreferencesHelper(context)
        return preferences.launchCount
    }

    @JvmStatic
    fun isDoNotShowAgain(context: Context): Boolean {
        val preferences = PreferencesHelper(context)
        return preferences.isDoNoShowAgain
    }

    @JvmStatic
    fun isTimeToRate(context: Context): Boolean {
        val resources = context.resources
        val dayUntilPrompt = resources.getInteger(R.integer.ar_app_rater_days_until_prompt)
        val launchesUntilPrompt = resources.getInteger(R.integer.ar_app_rater_launches_until_prompt)

        val preferences = PreferencesHelper(context)
        val launchCount = preferences.launchCount
        val firstLaunchDate = preferences.firstLaunchDate

        return !preferences.isDoNoShowAgain &&
            launchCount >= launchesUntilPrompt &&
            System.currentTimeMillis() >= firstLaunchDate + dayUntilPrompt * DAY
    }

    /**
     * Open Play Store for the related application.
     */
    fun rateApp(context: Context) {
        PreferencesHelper(context)
            .setDoNoShowAgain(true)

        // create Play Store intent
        var rate = Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + context.packageName))
        if (!isIntentCallable(context, rate)) {
            // if Play Store doesn't exist show in browser
            rate = Intent(
                Intent.ACTION_VIEW,
                Uri.parse("http://play.google.com/store/apps/details?id=" + context.packageName)
            )
        }

        if (!isIntentCallable(context, rate)) {
            // if browser app doesn't exist, show toast and quit
            Toast.makeText(context, context.getString(R.string.ar_unable_to_find_google_play), Toast.LENGTH_LONG).show()
            return
        }

        if (context is Activity) {
            context.startActivity(rate)
        } else {
            val intent = Intent.createChooser(rate, null)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            context.startActivity(intent)
        }
    }

    /**
     * Reset launch counter.
     */
    fun remindLater(context: Context) {
        PreferencesHelper(context)
            .setLaunchCount(0)
            .setFirstLaunchDate(System.currentTimeMillis())
    }

    /**
     * Cancel reminders.
     */
    fun cancelReminders(context: Context) {
        PreferencesHelper(context)
            .setDoNoShowAgain(true)
    }

    /**
     * Check if this is a first launch
     */
    fun isFirstLaunch(context: Context): Boolean {
        return PreferencesHelper(context)
            .firstLaunchDate == 0L
    }

    private fun isIntentCallable(context: Context, intent: Intent): Boolean {
        val list = context.packageManager
            .queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)
        return list.size > 0
    }
}
